endpoints/events/vtrack.go

```diff
@@ -0,0 +1,306 @@
+package events
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/golang/glog"
+	"github.com/julienschmidt/httprouter"
+	accountService "github.com/prebid/prebid-server/account"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/analytics"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/prebid_cache_client"
+	"github.com/prebid/prebid-server/stored_requests"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"sort"
+	"strings"
+	"time"
+)
+
+const (
+	AccountParameter   = "a"
+	ImpressionCloseTag = "</Impression>"
+	ImpressionOpenTag  = "<Impression>"
+)
+
+type vtrackEndpoint struct {
+	Cfg         *config.Configuration
+	Accounts    stored_requests.AccountFetcher
+	BidderInfos adapters.BidderInfos
+	Cache       prebid_cache_client.Client
+}
+
+type BidCacheRequest struct {
+	Puts []prebid_cache_client.Cacheable `json:"puts"`
+}
+
+type BidCacheResponse struct {
+	Responses []CacheObject `json:"responses"`
+}
+
+type CacheObject struct {
+	UUID string `json:"uuid"`
+}
+
+func NewVTrackEndpoint(cfg *config.Configuration, accounts stored_requests.AccountFetcher, cache prebid_cache_client.Client, bidderInfos adapters.BidderInfos) httprouter.Handle {
+	vte := &vtrackEndpoint{
+		Cfg:         cfg,
+		Accounts:    accounts,
+		BidderInfos: bidderInfos,
+		Cache:       cache,
+	}
+
+	return vte.Handle
+}
+
+// /vtrack Handler
+func (v *vtrackEndpoint) Handle(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+
+	// get account id from request parameter
+	accountId := getAccountId(r)
+
+	// account id is required
+	if accountId == "" {
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte(fmt.Sprintf("Account '%s' is required query parameter and can't be empty", AccountParameter)))
+		return
+	}
+
+	// parse puts request from request body
+	req, err := ParseVTrackRequest(r, v.Cfg.MaxRequestSize+1)
+
+	// check if there was any error while parsing puts request
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte(fmt.Sprintf("Invalid request: %s\n", err.Error())))
+		return
+	}
+
+	ctx := context.Background()
```

Comment:
We should probably add a configurable timeout of some kind for account lookup. This can be a follow-up PR in our backlog or for you to tackle if you'd like. Please advise.

Reply:
I agree, it's good practice - Refactored. 
It uses the TimeoutMS configuration variable to apply an execution time constraint to the endpoint's handler.